File: core/llm/stream_filter.py
"""Strip tool-call mechanics from a streamed delta sequence so the UI shows
prose only. Detects three shapes:

- `<tool_call>...</tool_call>` tags.
- Bare top-level JSON objects whose keys include `name`, `function_name`,
  or `tool` (the qwen2.5-coder bare-emission patterns).
- The same JSON wrapped in a markdown fence (```json ... ```), fence included.

Non-tool-call JSON (e.g. a snippet the model quotes in prose) is held while
the braces balance, then emitted in one chunk; the same goes for fenced JSON.
Lives in the same family as `tool_call_recovery` — both compensate for
Ollama's streaming tool-call gaps on qwen2.5-coder.
"""
from core.llm.json_repair import loads_or_repair

_NAME_KEYS = ("name", "function_name", "tool")
_TAG_OPEN = "<tool_call>"
_TAG_CLOSE = "</tool_call>"
_FENCE_LANG_CHARS = "_-+."
_WHITESPACE = " \t\r\n"


class StreamFilter:
    def __init__(self) -> None:
        self.mode: str = "prose"
        self.pending: str = ""
        self.depth: int = 0
        self.in_string: bool = False
        self.escape: bool = False
        self.json_start: int = 0
        self.json_end: int = 0

    def push(self, delta: str) -> str:
        out: list[str] = []
        for ch in delta:
            if self.mode == "prose":
                self._handle_prose(ch, out)
            elif self.mode == "tag":
                self.pending += ch
                if self.pending.endswith(_TAG_CLOSE):
                    self.pending = ""
                    self.mode = "prose"
            elif self.mode == "json":
                self.pending += ch
                self._advance_json(ch)
                if self.depth == 0:
                    self._close_json(out)
            elif self.mode == "fence_open":
                self._handle_fence_open(ch, out)
            elif self.mode == "fence_body":
                self._handle_fence_body(ch, out)
            elif self.mode == "fence_json":
                self.pending += ch
                self._advance_json(ch)
                if self.depth == 0:
                    self.json_end = len(self.pending)
                    self.mode = "fence_close"
            elif self.mode == "fence_close":
                self._handle_fence_close(ch, out)
        return "".join(out)

    def flush(self) -> str:
        """End-of-stream tail. Returns any held prose; drops partial tool-call
        buffers. A complete fenced JSON that is NOT a tool call is returned."""
        pending, mode = self.pending, self.mode
        self.pending = ""
        self.mode = "prose"
        if mode == "prose":
            return pending
        if mode == "fence_close" and not self._is_tool_call(pending[self.json_start:self.json_end]):
            return pending
        return ""

    def _handle_prose(self, ch: str, out: list[str]) -> None:
        if self.pending.startswith("<"):
            candidate = self.pending + ch
            if candidate == _TAG_OPEN:
                self.mode = "tag"
                self.pending = ""
                return
            if _TAG_OPEN.startswith(candidate):
                self.pending = candidate
                return
            out.append(self.pending)
            self.pending = ""
            self._handle_prose_fresh(ch, out)
            return
        self._handle_prose_fresh(ch, out)

    def _handle_prose_fresh(self, ch: str, out: list[str]) -> None:
        if ch == "<":
            self.pending = "<"
            return
        if ch == "{":
            self.mode = "json"
            self.pending = "{"
            self._reset_json_tracking()
            return
        if ch == "`":
            self.mode = "fence_open"
            self.pending = "`"
            return
        out.append(ch)

    # ---------------------------------------------------------------- fences

    def _handle_fence_open(self, ch: str, out: list[str]) -> None:
        """Accumulating ``` + optional language tag, until the newline."""
        only_ticks = not self.pending.strip("`")
        if ch == "`":
            if only_ticks and len(self.pending) < 3:
                self.pending += ch
                return
            self._abort_fence(ch, out)
            return
        if only_ticks and len(self.pending) < 3:
            self._abort_fence(ch, out)  # one or two backticks: inline code, not a fence
            return
        if ch == "\n":
            self.pending += ch
            self.mode = "fence_body"
            return
        if ch.isalnum() or ch in _FENCE_LANG_CHARS:
            self.pending += ch
            return
        self._abort_fence(ch, out)

    def _handle_fence_body(self, ch: str, out: list[str]) -> None:
        """Inside the fence, before the first non-whitespace character."""
        if ch in _WHITESPACE:
            self.pending += ch
            return
        if ch == "{":
            self.json_start = len(self.pending)
            self.pending += ch
            self._reset_json_tracking()
            self.mode = "fence_json"
            return
        self._abort_fence(ch, out)  # fence body isn't JSON — stream it normally

    def _handle_fence_close(self, ch: str, out: list[str]) -> None:
        """JSON balanced; expecting optional whitespace then the closing ```."""
        if ch in _WHITESPACE and not self.pending.endswith("`"):
            self.pending += ch
            return
        if ch == "`":
            self.pending += ch
            if self.pending.endswith("```"):
                if not self._is_tool_call(self.pending[self.json_start:self.json_end]):
                    out.append(self.pending)
                self.pending = ""
                self.mode = "prose"
            return
        self._abort_fence(ch, out)

    def _abort_fence(self, ch: str, out: list[str]) -> None:
        """Held text turned out not to be a tool-call fence: emit and resume prose."""
        out.append(self.pending)
        self.pending = ""
        self.mode = "prose"
        self._handle_prose_fresh(ch, out)

    # ------------------------------------------------------------------ json

    def _reset_json_tracking(self) -> None:
        self.depth = 1
        self.in_string = False
        self.escape = False

    def _advance_json(self, ch: str) -> None:
        if self.escape:
            self.escape = False
            return
        if self.in_string:
            if ch == "\\":
                self.escape = True
            elif ch == '"':
                self.in_string = False
            return
        if ch == '"':
            self.in_string = True
        elif ch == "{":
            self.depth += 1
        elif ch == "}":
            self.depth -= 1

    def _close_json(self, out: list[str]) -> None:
        if not self._is_tool_call(self.pending):
            out.append(self.pending)
        self.pending = ""
        self.mode = "prose"

    @staticmethod
    def _is_tool_call(text: str) -> bool:
        obj = loads_or_repair(text)
        return isinstance(obj, dict) and any(k in obj for k in _NAME_KEYS)

File: src/core/llm/stream-filter.ts
// Strip tool-call mechanics from a streamed delta sequence so the UI shows prose only.
// Port of core/llm/stream_filter.py. Detects three shapes the model sometimes emits as plain
// text instead of structured tool_calls:
//
//   - `<tool_call> ... </tool_call>` tag-wrapped blocks.
//   - Bare top-level JSON objects whose keys include `name`, `function_name`, or `tool`
//     (the qwen2.5-coder bare-emission patterns).
//   - The same JSON wrapped in a ```json ... ``` fence (fence included).
//
// Non-tool-call JSON (e.g. a snippet the model quotes in prose) is held while the braces
// balance, then emitted in one chunk; likewise for fenced JSON. Display-only: the structured
// message.tool_calls is captured separately and is unaffected by filtering.

import { isAlnum } from './is-alnum.js';
import { isRecord } from './is-record.js';
import { loadsOrRepair } from './loads-or-repair.js';

type FilterMode =
  | 'prose'
  | 'tag'
  | 'json'
  | 'fence_open'
  | 'fence_body'
  | 'fence_json'
  | 'fence_close';

const NAME_KEYS = ['name', 'function_name', 'tool'] as const;
const TAG_OPEN = '<tool_call>';
const TAG_CLOSE = '</tool_call>';
const FENCE_LANG_CHARS = '_-+.';
const WHITESPACE = ' \t\r\n';

export class StreamFilter {
  private mode: FilterMode = 'prose';
  private pending = '';
  private depth = 0;
  private inString = false;
  private escape = false;
  private jsonStart = 0;
  private jsonEnd = 0;

  /** Feed a raw delta; returns the visible (filtered) prose to display, possibly empty. */
  push(delta: string): string {
    const out: string[] = [];
    for (const ch of delta) {
      switch (this.mode) {
        case 'prose':
          this.handleProse(ch, out);
          break;
        case 'tag':
          this.pending += ch;
          if (this.pending.endsWith(TAG_CLOSE)) {
            this.pending = '';
            this.mode = 'prose';
          }
          break;
        case 'json':
          this.pending += ch;
          this.advanceJson(ch);
          if (this.depth === 0) this.closeJson(out);
          break;
        case 'fence_open':
          this.handleFenceOpen(ch, out);
          break;
        case 'fence_body':
          this.handleFenceBody(ch, out);
          break;
        case 'fence_json':
          this.pending += ch;
          this.advanceJson(ch);
          if (this.depth === 0) {
            this.jsonEnd = this.pending.length;
            this.mode = 'fence_close';
          }
          break;
        case 'fence_close':
          this.handleFenceClose(ch, out);
          break;
      }
    }
    return out.join('');
  }

  /**
   * End-of-stream tail. Returns any held prose; drops partial tool-call buffers. A complete
   * fenced JSON that is NOT a tool call is returned.
   */
  flush(): string {
    const { pending, mode } = this;
    this.pending = '';
    this.mode = 'prose';
    if (mode === 'prose') return pending;
    if (mode === 'fence_close' && !this.isToolCall(pending.slice(this.jsonStart, this.jsonEnd))) {
      return pending;
    }
    return '';
  }

  private handleProse(ch: string, out: string[]): void {
    if (this.pending.startsWith('<')) {
      const candidate = this.pending + ch;
      if (candidate === TAG_OPEN) {
        this.mode = 'tag';
        this.pending = '';
        return;
      }
      if (TAG_OPEN.startsWith(candidate)) {
        this.pending = candidate;
        return;
      }
      out.push(this.pending);
      this.pending = '';
      this.handleProseFresh(ch, out);
      return;
    }
    this.handleProseFresh(ch, out);
  }

  private handleProseFresh(ch: string, out: string[]): void {
    if (ch === '<') {
      this.pending = '<';
      return;
    }
    if (ch === '{') {
      this.mode = 'json';
      this.pending = '{';
      this.resetJsonTracking();
      return;
    }
    if (ch === '`') {
      this.mode = 'fence_open';
      this.pending = '`';
      return;
    }
    out.push(ch);
  }

  // ------------------------------------------------------------------ fences

  /** Accumulating ``` + optional language tag, until the newline. */
  private handleFenceOpen(ch: string, out: string[]): void {
    const onlyTicks = this.pending.replace(/^`+|`+$/g, '') === '';
    if (ch === '`') {
      if (onlyTicks && this.pending.length < 3) {
        this.pending += ch;
        return;
      }
      this.abortFence(ch, out);
      return;
    }
    if (onlyTicks && this.pending.length < 3) {
      // one or two backticks: inline code, not a fence
      this.abortFence(ch, out);
      return;
    }
    if (ch === '\n') {
      this.pending += ch;
      this.mode = 'fence_body';
      return;
    }
    if (isAlnum(ch) || FENCE_LANG_CHARS.includes(ch)) {
      this.pending += ch;
      return;
    }
    this.abortFence(ch, out);
  }

  /** Inside the fence, before the first non-whitespace character. */
  private handleFenceBody(ch: string, out: string[]): void {
    if (WHITESPACE.includes(ch)) {
      this.pending += ch;
      return;
    }
    if (ch === '{') {
      this.jsonStart = this.pending.length;
      this.pending += ch;
      this.resetJsonTracking();
      this.mode = 'fence_json';
      return;
    }
    // fence body isn't JSON — stream it normally
    this.abortFence(ch, out);
  }

  /** JSON balanced; expecting optional whitespace then the closing ```. */
  private handleFenceClose(ch: string, out: string[]): void {
    if (WHITESPACE.includes(ch) && !this.pending.endsWith('`')) {
      this.pending += ch;
      return;
    }
    if (ch === '`') {
      this.pending += ch;
      if (this.pending.endsWith('```')) {
        if (!this.isToolCall(this.pending.slice(this.jsonStart, this.jsonEnd))) {
          out.push(this.pending);
        }
        this.pending = '';
        this.mode = 'prose';
      }
      return;
    }
    this.abortFence(ch, out);
  }

  /** Held text turned out not to be a tool-call fence: emit and resume prose. */
  private abortFence(ch: string, out: string[]): void {
    out.push(this.pending);
    this.pending = '';
    this.mode = 'prose';
    this.handleProseFresh(ch, out);
  }

  // -------------------------------------------------------------------- json

  private resetJsonTracking(): void {
    this.depth = 1;
    this.inString = false;
    this.escape = false;
  }

  private advanceJson(ch: string): void {
    if (this.escape) {
      this.escape = false;
      return;
    }
    if (this.inString) {
      if (ch === '\\') {
        this.escape = true;
      } else if (ch === '"') {
        this.inString = false;
      }
      return;
    }
    if (ch === '"') {
      this.inString = true;
    } else if (ch === '{') {
      this.depth += 1;
    } else if (ch === '}') {
      this.depth -= 1;
    }
  }

  private closeJson(out: string[]): void {
    if (!this.isToolCall(this.pending)) {
      out.push(this.pending);
    }
    this.pending = '';
    this.mode = 'prose';
  }

  private isToolCall(text: string): boolean {
    // loadsOrRepair: JSON.parse, falling back to a pass that escapes the literal control characters
    // qwen writes inside strings. isRecord rejects null and arrays before the `in` probes below.
    const obj = loadsOrRepair(text);
    return isRecord(obj) && NAME_KEYS.some((k) => k in obj);
  }
}
